src/tactic/lift.lean

```diff
@@ -0,0 +1,128 @@
+/-
+Copyright (c) 2019 Floris van Doorn. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Floris van Doorn
+-/
+import tactic.rcases
+
+/-!
+# lift tactic
+
+This file defines the lift tactic, allowing the user to lift elements from one type to another
+under a specified condition.
+
+## Tags
+
+lift, tactic
+-/
+
+universe variables u v
+
+run_cmd mk_simp_attr `can_lift
+
+/-- A class specifying that you can lift elements from `α` to `β` assuming `cond` is true.
+  Used by the tactic `lift`. For the proper functioning of the lift tactic, make sure that all
+  instances of this class have attribute `[can_lift]`. -/
+class can_lift (α : Type u) (β : Type v) : Type (max u v) :=
+(coe : β → α)
+(cond : α → Prop)
+(prf : ∀(x : α), cond x → ∃(y : β), coe y = x)
+
+@[can_lift] instance : can_lift ℤ ℕ :=
+⟨coe, λ n, 0 ≤ n, λ n hn, ⟨n.nat_abs, int.nat_abs_of_nonneg hn⟩⟩
+
+namespace tactic
+
+/- Construct the proof of `cond x` in the lift tactic.
+  If the proof was specified, we check whether it has the correct type.
+    If it doesn't have the correct type, we display an error message
+    (but first call dsimp on the expression in the message).
+  If the proof was not specified, we create assert it as a local constant.
+  (The name of this local constant doesn't matter, since `lift` will remove it from the context) -/
+meta def get_lift_prf (h : option pexpr) (old_tp new_tp inst e : expr) : tactic expr :=
```

Comment:
The doc string should explicitly say what each of these arguments are. Between the names and description, I can mostly figure it out, but it takes some thought.

Reply:
done